commandapi-core/src/main/java/dev/jorel/commandapi/network/CommandAPIProtocol.java

```diff
@@ -0,0 +1,186 @@
+package dev.jorel.commandapi.network;
+
+import dev.jorel.commandapi.network.packets.ClientToServerUpdateRequirementsPacket;
+import dev.jorel.commandapi.network.packets.ProtocolVersionTooOldPacket;
+import dev.jorel.commandapi.network.packets.SetVersionPacket;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.function.Function;
+
+// Inspired by net.minecraft.network.ConnectionProtocol
+/**
+ * An enum that defines the different channels of the network protocol of the CommandAPI. It handles converting bytes off
+ * the network into {@link CommandAPIPacket}s.
+ */
+public enum CommandAPIProtocol {
+	// Protocol states
+	/**
+	 * A {@link CommandAPIProtocol} for the {@code commandapi:handshake} channel. The packets in this protocol help instances
+	 * of the CommandAPI communicate the network data they understand.
+	 */
+	HANDSHAKE("commandapi:handshake", new PacketSetBuilder()
+		.register(SetVersionPacket.class, SetVersionPacket::deserialize)
+		.register(ProtocolVersionTooOldPacket.class, ProtocolVersionTooOldPacket::deserialize)
+	),
+	/**
+	 * A {@link CommandAPIProtocol} for the {@code commandapi:play} channel. The packets in this protocol handle events
+	 * that happen while a server is running.
+	 */
+	PLAY("commandapi:play", new PacketSetBuilder()
+		.register(ClientToServerUpdateRequirementsPacket.class, ClientToServerUpdateRequirementsPacket::deserialize)
+	);
+
+	// Global channel variables
+	// TODO: If the first released version of the CommandAPI that can receive messages is not 9.0.4, change this description
+	/**
+	 * The current version of the protocol. This should be incremented when the protocol is updated to indicate a large
+	 * change. A connection without a communicating instance of the CommandAPI (either no CommandAPI instance or a version
+	 * before 9.0.4) should be treated as protocol version 0.
```

Comment:
Other things can cause the protocol version to be updated. Said another way,

- Any change to the data that gets sent over the network should get a new protocol version. 

OR

- Any change that would confuse a previous protocol version.

For example:

Adding packet `A`
- The data that can be sent has changed, since the new id for packet `A` may be sent when it wasn't there before. Version 1 would be confused since it doesn't recognize the new id.

Adding new data to an existing packet. Say, add a String field for the name of the platform to `SetVersionPacket`
- The data that can be sent has changed, since there is now a String at the end of the `SetVersionPacket`. Version 1 would be confused since it doesn't expect extra data in the `SetVersionPacket`.

Reply:
That makes sense, thanks!